lib/msal-node/src/network/LoopbackClient.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { Constants as CommonConstants, ServerAuthorizationCodeResponse, UrlString } from "@azure/msal-common";
+import { createServer, IncomingMessage, Server, ServerResponse } from "http";
+import { NodeAuthError } from "../error/NodeAuthError";
+import { Constants, LOOPBACK_SERVER_CONSTANTS } from "../utils/Constants";
+
+export class LoopbackClient {
+    private server: Server;
+
+    /**
+     * Spins up a loopback server which returns the server response when the localhost redirectUri is hit
+     * @param successTemplate 
+     * @param errorTemplate 
+     * @returns 
+     */
+    async listenForAuthCode(successTemplate?: string, errorTemplate?: string): Promise<ServerAuthorizationCodeResponse> {
+        if (!!this.server) {
+            throw NodeAuthError.createLoopbackServerAlreadyExistsError();
+        }
+
+        return new Promise((resolve, reject) => {
+            this.server = createServer(async (req: IncomingMessage, res: ServerResponse) => {
+                const url = req.url;
+                if (!url) {
+                    res.end(errorTemplate || "Error occurred loading redirectUrl");
+                    reject(NodeAuthError.createUnableToLoadRedirectUrlError());
+                    return;
+                } else if (url === CommonConstants.FORWARD_SLASH) {
+                    res.end(successTemplate || "Auth code was successfully acquired. You can close this window now.");
+                    return;
+                }
+    
+                const authCodeResponse = UrlString.getDeserializedQueryString(url);
+                if (authCodeResponse.code) {
+                    const redirectUri = await this.getRedirectUri();
+                    res.writeHead(302, { location: redirectUri }); // Prevent auth code from being saved in the browser history
+                    res.end();
+                }
+                resolve(authCodeResponse);
+            });
+            this.server.listen(0); // Listen on any available port
+        });
+    }
+
+    /**
+     * Get the port that the loopback server is running on
+     * @returns 
+     */
+    async getRedirectUri(): Promise<string> {
+        if (!this.server) {
+            throw NodeAuthError.createNoLoopbackServerExistsError();
+        }
+        const port: number = await new Promise((resolve, reject) => {
+            let ticks = 0;
+            const id = setInterval(() => {
```

Comment:
In reality this should return on the first or second interval (100ms intervals for 5s were just the defaults I chose arbitrarily, we probably could get away with a much shorter timeout). The reason this is needed is because the server's event listener may not be registered yet by the time this gets called, resulting in a null response. 

An alternative solution (which is perhaps more intuitive) would have been to use the interval to check if the server is listening and then get the port once as you expect. Since the interval is needed either way I took the shortcut - though I see that was maybe not the right choice for readability. I'm open to reconsidering that decision.

Reply:
Can you add a comment detailing this implementation detail?

Reply:
> In reality this should return on the first or second interval (100ms intervals for 5s were just the defaults I chose arbitrarily, we probably could get away with a much shorter timeout). The reason this is needed is because the server's event listener may not be registered yet by the time this gets called, resulting in a null response.
> 
> An alternative solution (which is perhaps more intuitive) would have been to use the interval to check if the server is listening and then get the port once as you expect. Since the interval is needed either way I took the shortcut - though I see that was maybe not the right choice for readability. I'm open to reconsidering that decision.

Sounds like polling and I would prefer the separation. Even if technically what you did works, it would be easier to maintain and improve the code if we separate the listening and port acquisition parts, even for performance measurements and telemetry.

Reply:
Either a comment on the current approach as Jason suggested or a separation as Sameera suggested would be OK here.  I'd lean more towards Sameera's suggestion as that implementation approach feels cleaner.

Reply:
Done - refactored to poll for the listener after creating it.